query-parser/src/util.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+extern crate edn;
+extern crate mentat_query;
+
+use std::collections::BTreeMap;
+
+use self::edn::Value::PlainSymbol;
+use self::mentat_query::Variable;
+
+use super::error::{FindParseError};
+
+/// If the provided EDN value is a PlainSymbol beginning with '?', return
+/// it wrapped in a Variable. If not, return None.
+pub fn value_to_variable(v: &edn::Value) -> Option<Variable> {
+    if let PlainSymbol(ref sym) = *v {
+        if sym.0.starts_with('?') {
+            return Some(Variable(sym.clone()));
+        }
+    }
+    return None;
+}
+
+/// If the provided slice of EDN values are all variables as
+/// defined by `value_to_variable`, return a Vec of Variables.
+/// Otherwise, return an error.
+pub fn values_to_variables(vals: &[edn::Value]) -> Result<Vec<Variable>, FindParseError> {
+    let mut out: Vec<Variable> = Vec::with_capacity(vals.len());
```

Comment:
You can do this less explicitly with `map`, `collect`, and `ok_or`.  It's worth internalizing that there is a folding/aborting `collect` implementation for `[Option<T>] -> Option<Vec<T>>` and similarly for `Result`.